torch/nn/modules/module.py

```diff
@@ -2007,7 +2018,9 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
                 :meth:`~torch.nn.Module.load_state_dict`
         """
         for hook in self._load_state_dict_pre_hooks.values():
-            hook(state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs)
+            hook_result = hook(state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs)
+            if hook_result is not None:
```

Comment:
What does it mean for this hook to return a new state_dict? Why is the situation different from the other hook where we removed this behavior?

Reply:
This was a request from @janeyx99 in https://github.com/pytorch/pytorch/issues/75287 for consistency with optimizer

> EDIT (@janeyx99): We should also allow the hook to optionally return a new state_dict, if desired, as it will cover the use case where one may want to completely change up the state_dict. Relevant PR https://github.com/pytorch/pytorch/pull/105953 where we have this for the optimizer.

I think the argument there was that it is a convenience if say users want to pass an entirely new state_dict in their hook. Since the `state_dict` that is passed into `_load_from_state_dict` is a dictionary containing the keys for the submodule (and its recursive children), the effects of the hook called by the child don't affect the parent, so the child doesn't have to be "aware" of the parent

But tbh I think it makes sense to remove this for consistency with the state_dict hook like you mentioned.

Hm, in the `state_dict` case, the result from the module is always being passed to the children's recursive state_dict calls, 

https://github.com/pytorch/pytorch/blob/04272a0e129a7314275d21f871e5fcadc37c796d/torch/nn/modules/module.py#L1930-L1980

So let's say child's hook is out of place like `return {k: v+2 for k, v in m.parameters()}`, then if the out of place behavior of the child's hook was respected (which it currently isn't), it would delete all the keys of the parent from the state_dict is returned, so every hook that returned something out of place would need to be written in a "parent-aware" manner.

Reply:
Removed this behavior